app/models/package_manager/go.rb

```diff
@@ -53,34 +53,54 @@ def self.update(name)
       # example: github.com/myexample/modulename/v2
       if (matches = name.match(VERSION_MODULE_REGEX))
         PackageManagerDownloadWorker.perform_async(self.name, matches[1])
+
+        module_project = Project.find_by(platform: "Go", name: name)
+        base_module_project = Project.find_by(platform: "Go", name: matches[1])
+        return if module_project.nil? || base_module_project.nil?
+
+        module_project.versions.each do |version|
+          base_module_project.versions.find_or_create_by(number: version[:number]) do |new_version|
+            new_version.update(published_at: version[:published_at])
+            new_version.save
+          end
+        end
       end
     end
 
     def self.project(name)
       if (doc_html = get_html("https://pkg.go.dev/#{name}"))
-        # NB fetching versions from the html only gets dates without timestamps, but we could alternatively use the go proxy too:
-        #   1) Fetch the list of versions: https://proxy.golang.org/#{module_name}/@v/list
-        #   2) And for each version, fetch https://proxy.golang.org/#{module_name}/@v/#{v}.info
-        versions_html = get_html("https://pkg.go.dev/#{name}?tab=versions")
-
-        { name: name, html: doc_html, overview_html: doc_html, versions_html: versions_html }
+        { name: name, html: doc_html, overview_html: doc_html }
       else
         { name: name }
       end
     end
 
     def self.versions(project, _name)
       return [] if project.nil?
-
-      versions = project[:versions_html]&.css(".Versions-item")&.map do |v|
-        { number: v.css("a").first.text, published_at: Chronic.parse(v.css(".Versions-commitTime").first.text) }
-      end
-
-      if (matches = project[:name].match(VERSION_MODULE_REGEX))
-        versions = versions.select { |version| version[:number].start_with?(matches[2]) }
-      end
-
-      versions
+      return project[:versions] if project[:versions]
+
+      known_versions = Project.find_by(platform: "Go", name: project[:name])&.versions
+
+      # NB fetching versions from the html only gets dates without timestamps, but we could alternatively use the go proxy too:
+      #   1) Fetch the list of versions: https://proxy.golang.org/#{module_name}/@v/list
+      #   2) And for each version, fetch https://proxy.golang.org/#{module_name}/@v/#{v}.info
+
+      get_raw("http://proxy.golang.org/#{project[:name]}/@v/list")
+        &.lines
+        &.map(&:strip)
+        &.reject(&:blank?)
+        &.map do |v|
+          return known_versions.find_by(number: v).slice(:number, :published_at) if known_versions&.find_by(number: v).present?
```

Comment:
Followup to my last comment, this could then be:

```suggestion
          known_version = known_versions.find { |(number, _published_at)| number == v }
          return known_version.slice(:number, :published_at) if known_version.present?
```